File: packages/project-ops/src/commands.ts
import { z } from "zod";
import { GridSchema, ImageRefSchema, LinkSchema, InlineLocationContentSchema, FogOfWarSchema, type HexenProject } from "@hex-enductor/hexen-schema";
import { saveLink, saveLocationContent, addLocationLink, saveGrid, saveImage, setFog, setFogCells } from "./mutations.ts";

const LinkPatchSchema = LinkSchema.partial();
const InlineContentPatchSchema = InlineLocationContentSchema.omit({ type: true }).partial();

export const SaveLinkCommandSchema = z.object({
  type: z.literal("saveLink"),
  locationId: z.string(),
  linkId: z.string(),
  patch: LinkPatchSchema,
});

export const SaveLocationContentCommandSchema = z.object({
  type: z.literal("saveLocationContent"),
  locationId: z.string(),
  patch: InlineContentPatchSchema,
});

// The pin's own marker type (settlement/ruin/landmark/...) is named
// `linkType` here, not `type` — `type` is already the command envelope's
// own discriminant.
export const AddLocationLinkCommandSchema = z.object({
  type: z.literal("addLocationLink"),
  parentLocationId: z.string(),
  targetLocationId: z.string().min(1),
  x: z.number(),
  y: z.number(),
  linkType: z.string().min(1),
});

export const SaveGridCommandSchema = z.object({
  type: z.literal("saveGrid"),
  locationId: z.string(),
  grid: GridSchema.nullable(),
});

export const SaveImageCommandSchema = z.object({
  type: z.literal("saveImage"),
  locationId: z.string(),
  image: ImageRefSchema.nullable(),
});

export const SetFogCommandSchema = z.object({
  type: z.literal("setFog"),
  locationId: z.string(),
  fog: FogOfWarSchema.nullable(),
});

export const SetFogCellsCommandSchema = z.object({
  type: z.literal("setFogCells"),
  locationId: z.string(),
  cells: z.array(z.string()),
  revealed: z.boolean(),
});

/**
 * Every interaction that mutates a HexenProject, as data — the same
 * shape whether it arrives over hexend's WS session or is applied
 * straight to an in-memory project by the local-fs storage backend.
 * Keeping the union and its reducer here (next to the pure mutation
 * functions) means both transports validate and apply commands
 * identically.
 */
export const CommandSchema = z.discriminatedUnion("type", [
  SaveLinkCommandSchema,
  SaveLocationContentCommandSchema,
  AddLocationLinkCommandSchema,
  SaveGridCommandSchema,
  SaveImageCommandSchema,
  SetFogCommandSchema,
  SetFogCellsCommandSchema,
]);
export type Command = z.infer<typeof CommandSchema>;

export function applyCommand(project: HexenProject, command: Command): HexenProject {
  switch (command.type) {
    case "saveLink":
      return saveLink(project, command.locationId, command.linkId, command.patch);
    case "saveLocationContent":
      return saveLocationContent(project, command.locationId, command.patch);
    case "addLocationLink":
      return addLocationLink(
        project,
        command.parentLocationId,
        command.targetLocationId,
        command.x,
        command.y,
        command.linkType,
      );
    case "saveGrid":
      return saveGrid(project, command.locationId, command.grid);
    case "saveImage":
      return saveImage(project, command.locationId, command.image);
    case "setFog":
      return setFog(project, command.locationId, command.fog);
    case "setFogCells":
      return setFogCells(project, command.locationId, command.cells, command.revealed);
  }
}

File: apps/hexend/src/commands.rs
//! The Command union's reducer. Port of packages/project-ops/src/commands.ts.

use crate::mutations::{
    add_location, add_location_link, save_grid, save_image, save_link, save_location_content,
    set_fog, set_fog_cells, MutationError,
};
use crate::pb::hexen::v1::{command, Command, HexenProject};

pub fn apply_command(project: &mut HexenProject, command: Command) -> Result<(), MutationError> {
    match command.kind {
        Some(command::Kind::SaveLink(c)) => {
            save_link(project, &c.location_id, &c.link_id, c.patch.unwrap_or_default())
        }
        Some(command::Kind::SaveLocationContent(c)) => {
            save_location_content(project, &c.location_id, c.patch.unwrap_or_default())
        }
        Some(command::Kind::AddLocationLink(c)) => add_location_link(
            project,
            &c.parent_location_id,
            &c.target_location_id,
            c.x,
            c.y,
            c.link_type,
        ),
        Some(command::Kind::SaveGrid(c)) => save_grid(project, &c.location_id, c.grid),
        Some(command::Kind::SaveImage(c)) => save_image(project, &c.location_id, c.image),
        Some(command::Kind::SetFog(c)) => set_fog(project, &c.location_id, c.fog),
        Some(command::Kind::SetFogCells(c)) => set_fog_cells(project, &c.location_id, &c.cells, c.revealed),
        Some(command::Kind::AddLocation(c)) => add_location(project, &c.location_id),
        None => Err(MutationError::EmptyCommand),
    }
}
